### DISCUSSION TITLE
Allow dynamic list of labels in custom legend

### DISCUSSION THREAD

**Discussion by AntoineThebaud** [grafana/grafana#51386] [Dashboards, Panels & Visualizations]
(datasource: Prometheus)

It's not possible currently to provide a custom formatting of the legend when having a dynamic list of labels in the query result.

See an example with the following:
![image](https://user-images.githubusercontent.com/7058693/175548196-b45f422b-f627-4f27-80dd-2de72a1101af.png)

Here `{{label_name}}` actually equals to empty string, thus it falls back to the Auto legend,. So if I try to add some text behind, it stops being rendered:
![image](https://user-images.githubusercontent.com/7058693/175548993-74d19c08-1658-4c28-bdb9-5e950d8a7ab6.png)

For sure if I provide the actual labels being used, it works:
![image](https://user-images.githubusercontent.com/7058693/175549325-feaec860-1200-4904-9f35-7b25d427a90a.png)

But again, the list of labels is dynamically generated from the grouping variable here, so we dont want to 'hardcode' the labels like that.

We actually found a solution that works when selecting 1 value only in the `grouping` dropdown: "double interpolation" by interpolating the value of the dropdown within curly braces:
![image](https://user-images.githubusercontent.com/7058693/175549937-00ef08a4-78a4-4639-98bf-938188fa7621.png)

But for sure this doesnt work anymore when you select multiple values:
![image](https://user-images.githubusercontent.com/7058693/175550151-6acd099d-16c2-4e65-b8b8-f9406d683497.png)

Thus I am looking for a solution to achieve this. We thought about a solution with @w4rgrum that would be to add a new [variable format option](https://grafana.com/docs/grafana/latest/variables/advanced-variable-format-options/). This could look like this:

```typescript
{
  id: FormatRegistryID.legend,
  name: 'Legend',
  description: 'Values are formatted to be used as legend variables and multi-valued variables are separated by ","',
  formatter: ({ value }) => {
    if (typeof value === 'string') {
      return '{{ ' + value + ' }}';
    }
    if (value instanceof Array && value.length === 0) {
      return '__empty__';
    }
    return '{{ ' + value.join(' }},{{ ') + ' }}';
  },
},
```

For sure the separator character (here comma) is subject to debate.
So in the end we would input in the legend textbox: `${grouping:legend}-custom`

Does that proposal look fine to you? If not, how could the implementation be done to properly allow using a dynamic list of labels in a custom legend?

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #82998. Feel free to continue the discussion around this there. Thank you!
